debug: fix data race in Debug::log()

Daviey opened the pull request:
Fixes a data race in the logging system that causes crashes when plugins use background threads.

**The issue:**

The templated `Debug::log(eLogLevel, fmt, args...)` locks `m_logMutex` and calls the non-templated `Debug::log(eLogLevel, std::string)`, but the non-templated version doesn't lock anything. This causes crashes when background threads (like in hypr-dynamic-cursors) call `Debug::log()` while the main thread is also logging.

**The fix:**

Added a private `logImpl()` that does the actual logging work, with both public `log()` entry points handling the locking. This uses a regular mutex instead of recursive_mutex and avoids duplicate checks.

Fixes: #11929
Fixes: VirtCode/hypr-dynamic-cursors#99

See also: https://github.com/hyprwm/Hyprland/discussions/11930

vaxerski commented:
I mean, wouldn't that be a plugin bug?

VirtCode commented:
> I mean, wouldn't that be a plugin bug?

I think the logging is supposed to be thread-safe since https://github.com/hyprwm/Hyprland/commit/548968279926a73d7ff19a9a185c977c50d56756 which is why I thought I'd just use it. Yes I know you explicitly discourage using extra threads in plugins, but loading a large cursor theme is horrendously slow that it would otherwise increase compositor startup times with dynamic cursors by literal seconds.

The discrepancy between the templated logging function and the non-templated has been bothering me anyways since quite some time because if you have enabled timestamps in the log, the timestamp only gets introduced when using the templated version, and not the normal one. This means the log is pretty inconsistent, only introducing timestamps in some cases, like e.g.
```
[LOG] [CPrimarySelectionProtocol] CPrimarySelectionProtocol::updateSelection: cannot send selection to a client without a data_device
[LOG] [22:30:35.317512890] CWLSurface 5581a8441d30 called destroy()
[LOG] Cleanup: destroyed a layersurface
[LOG] [22:30:35.322383355] Framebuffer created, status 36053
```

So maybe we should do this in this PR too if we are already at it.

As for the fix in this PR, it would probably also be enough to just move the mutex locking from the templated version into the normal version, because actual writes which could cause issues only happen in there. The only thing the `logImpl` approach saves is the _second_ trace and shutdown check in case of the templated version, which I don't know whether that's worth an extra function call.

vaxerski commented:
Ah, right makes sense.

Daviey commented:
Updated to use the simpler approach you suggested. Removed the `logImpl()` function entirely and just moved the mutex lock from the template version into the non-template version.

Now the template version checks trace/shutdown conditions, formats the message (including timestamp if enabled), and calls the non-template version with the formatted string. The non-template version checks trace/shutdown conditions, acquires the mutex, and does all the actual writes to m_rollingLog, m_logOfs, etc.

This way we only lock once per log call, and all the shared state writes are protected. No need for the extra function or any duplicate checks.

@VirtCode - you're right that this is simpler and achieves the same thing. The timestamp inconsistency you mentioned is still there (template adds it, non-template doesn't), but that's a separate issue from the thread safety fix.

Thanks for the feedback, much cleaner this way.
